### PR TITLE
[semantic-arc-opts] Eliminate all dead live ranges.

### PR DESCRIPTION
**Pull request by gottesmm**
Previously, we only did single copy/destroys. Now we handle a copy + any number
of destroy_values.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp [hunk 1]
```diff
@@ -193,6 +194,39 @@ bool SemanticARCOptVisitor::visitCopyValueInst(CopyValueInst *cvi) {
...
+
+        // Stash dvi in destroys so we can easily eliminate it later and
+        // then...
+        destroys.push_back(dvi);
+        return true;
+      })) {
+    return false;
+  }
+
+  // Now that we have a truly dead live range copy value, eliminate it!x
```
Comment (shajrawi) [inline lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp]: extra x after the exclamation mark

File: lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp [hunk 2]
```diff
@@ -193,6 +194,39 @@ bool SemanticARCOptVisitor::visitCopyValueInst(CopyValueInst *cvi) {
...
+  // instructions. We begin by performing that check and gathering up our
+  // destroy_value.
+  SmallVector<DestroyValueInst *, 16> destroys;
+  if (!all_of(cvi->getUses(), [&](Operand *op) {
+        auto *dvi = dyn_cast<DestroyValueInst>(op->getUser());
+        if (!dvi)
+          return false;
+
+        // Stash dvi in destroys so we can easily eliminate it later and
+        // then...
```
Comment (shajrawi) [inline lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp]: Then... ? either remove the and then or complete the sentence :)

File: lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp [hunk 3]
```diff
@@ -193,6 +194,39 @@ bool SemanticARCOptVisitor::visitCopyValueInst(CopyValueInst *cvi) {
...
+    destroys.pop_back_val()->eraseFromParent();
+    ++NumEliminatedInsts;
+  }
+  cvi->eraseFromParent();
+  ++NumEliminatedInsts;
+  return true;
+}
+
+bool SemanticARCOptVisitor::visitCopyValueInst(CopyValueInst *cvi) {
+  // If our copy value inst has a only destroy_value users, it is a dead live
```
Comment (shajrawi) [inline lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp]: "has a only"? rephrase

File: test/SILOptimizer/semantic-arc-opts.sil [hunk 4]
```diff
@@ -188,3 +189,53 @@ bb0(%0 : @unowned $Builtin.NativeObject):
...
+bb2:
+  %2 = function_ref @owned_user : $@convention(thin) (@owned Builtin.NativeObject) -> ()
+  apply %2(%1) : $@convention(thin) (@owned Builtin.NativeObject) -> ()
+  br bb3
+
+bb3:
+  destroy_value %0 : $Builtin.NativeObject
+  %9999 = tuple()
+  return %9999 : $()
+}
```
Comment (shajrawi) [inline test/SILOptimizer/semantic-arc-opts.sil]: new line at end of file

**Review by shajrawi — COMMENTED** [review]
Looks good, but, please update the comments.

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
More out of curiosity. @swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
I am going to kill the test and merge job so that I can see the benchmark results.

**Comment by gottesmm** [comment]
The OS X build will still finish and be marked as success though.

**Comment by gottesmm** [comment]
Nope too slow.

**Pull request merged**